### PR TITLE
fix(learn): prevent indexing learn pages on search engines

### LINKED ISSUE #38164: Noindex /learn
As part of cleaning up our Google footprint, we want to de-index /learn. We don't want people to arrive from Google to the middle of a challenge (because they most likely won't be ready to sit down and code right then and there).

So we need to add this to all pages of /learn

```html
<meta name="robots" content="noindex">
```

### PR DESCRIPTION
**Pull request by hassaanp**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] All the files I changed are in the same world language, for example: only English changes, or only Chinese changes, etc.

Closes #38164

I have followed the [guidelines](https://support.google.com/webmasters/answer/6062608?hl=en) prescribed by Google that suggests using robot.txt and `noindex` meta tag to prevent a web page from being accessed by crawlers. Adding a robot.txt is not enough since it does not hide the page from the Google, it helps in managing the traffic. Therefore, I have used Helmet to add the `noindex` meta tag in the learn layout that is used by the curriculum challenges.

### CODE REVIEW

**Comment by raisedadead** [comment]
> Adding a robot.txt is not enough since it does not hide the page from the Google, it helps in managing the traffic.

Thanks for researching this bit.

File: config/env.js [hunk 1]
```diff
@@ -20,7 +20,8 @@ const {
   STRIPE_PUBLIC_KEY: stripePublicKey,
   SERVICEBOT_ID: servicebotId,
   ALGOLIA_APP_ID: algoliaAppId,
-  ALGOLIA_API_KEY: algoliaAPIKey
+  ALGOLIA_API_KEY: algoliaAPIKey,
+  SITE_URL: siteUrl
```
Comment (raisedadead) [inline config/env.js]:
```suggestion

```

This can go, we can re-use the `HOME_LOCATION` instead

File: config/env.js [hunk 2]
```diff
@@ -29,7 +30,8 @@ const locations = {
   forumLocation: forum,
   newsLocation: news,
   forumProxy: forumProxy,
-  newsProxy: newsProxy
+  newsProxy: newsProxy,
+  siteUrl: siteUrl
```
Comment (raisedadead) [inline config/env.js]:
```suggestion

```

This can go, we can re-use the `HOME_LOCATION` instead

File: sample.env [hunk 3]
```diff
@@ -43,3 +43,5 @@ TEST_CHALLENGES_FOR_LANGS=english
 DOCKER_HOST_LOCATION=localhost
 
 GHOST_CLIENT_KEY=123abc
+
+SITE_URL='https://www.freecodecamp.org'
```
Comment (raisedadead) [inline sample.env]:
```suggestion

```

This can go, we can re-use the `HOME_LOCATION` instead

File: client/gatsby-config.js [hunk 4]
```diff
@@ -11,7 +11,7 @@ const curriculumIntroRoot = path.resolve(__dirname, './src/pages');
 module.exports = {
   siteMetadata: {
     title: 'freeCodeCamp',
-    siteUrl: 'https://www.freecodecamp.org'
+    siteUrl: config.siteUrl
```
Comment (raisedadead) [inline client/gatsby-config.js]:
```suggestion
    siteUrl: config.homeLocation
```

We can re-use the `HOME_LOCATION` instead.

File: client/gatsby-config.js [hunk 5]
```diff
@@ -1,5 +1,5 @@
 const path = require('path');
-
+const config = require('./config/env');
```
Comment (raisedadead) [inline client/gatsby-config.js]:
```suggestion
const config = require('./config/env');

```

Nitpick on formatting.

**Review by raisedadead — CHANGES_REQUESTED** [review]
I think we are able to re-use the `HOME_LOCATION` instead of a new variable.

**Comment by raisedadead** [comment]
This should be ready to go once the CI passes. Can someone from @freeCodeCamp/dev-team check and merge, since I have a commit.

File: client/gatsby-config.js [hunk 6]
```diff
@@ -131,6 +132,12 @@ module.exports = {
...
       }
     },
-    'gatsby-plugin-remove-serviceworker'
+    'gatsby-plugin-remove-serviceworker',
+    {
+      resolve: 'gatsby-plugin-robots-txt',
+      options: {
+        policy: [{ userAgent: '*', disallow: '/learn', crawlDelay: 2 }]
+      }
+    }
```
Comment (ojeytonwilliams) [inline client/gatsby-config.js]:
```suggestion
```
I think we should be either using `robots.txt` or adding `noindex`, not both.  Looking [here](https://support.google.com/webmasters/answer/93710)  Google points out that the `robots.txt` will mean that the bot never goes into `/learn` and so never sees `noindex`.

File: client/src/components/layouts/Learn.js [hunk 7]
```diff
@@ -41,6 +41,13 @@ class LearnLayout extends Component {
     this.props.tryToShowDonationModal();
   }
 
+  componentWillUnmount() {
+    const metaTag = document.querySelector(`meta[name="googlebot"]`);
```
Comment (ojeytonwilliams) [inline client/src/components/layouts/Learn.js]:
```suggestion
    const metaTag = document.querySelector(`meta[name="robots"]`);
```
`robots` is more generic and includes `googlebot`

File: client/src/components/layouts/Learn.js [hunk 8]
```diff
@@ -59,6 +66,9 @@ class LearnLayout extends Component {
 
     return (
       <Fragment>
+        <Helmet>
+          <meta content='noindex' name='googlebot' />
```
Comment (ojeytonwilliams) [inline client/src/components/layouts/Learn.js]:
```suggestion
          <meta content='noindex' name='robots' />
```
See above.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
If we go the `noindex` route package.json and package-lock.json should be reverted.

File: client/gatsby-config.js (hunk 6 above)
Comment (hassaanp) [reply client/gatsby-config.js]: That is interesting. I did not know that. Thanks for sharing this. At another similar post from Google, I got the impression that perhaps both things needed to be done. Perhaps that was poor documentation on their part - never mind. Sure, let's go ahead with the `noindex` directive.

File: client/src/components/layouts/Learn.js (hunk 7 above)
Comment (hassaanp) [reply client/src/components/layouts/Learn.js]: this needs to be changed on two different lines. Pushing a commit with the complete changes.

**Comment by ojeytonwilliams** [comment]
@hassaanp something must have gone wrong with the merge - Mrugesh's commits were lost.  I had them locally, so I've restored them and applied the other changes too.

@ahmadabdolsaheb since I've *also* added commits, could you take a quick look and merge if you're happy?

**Review by ojeytonwilliams — APPROVED** [review]

**Pull request merged**